File: test/ShmemPrimitive_test.cpp
#include <gtest/gtest.h>
#include <spdlog/spdlog.h>
#include <spdlog/sinks/stdout_color_sinks.h>
#include <cstring>
#include <iostream>
#include "ShmemPrimitive.h"
#include "ShmemAccessor.h"
using namespace std;
namespace py = pybind11;

class ShmemPrimitiveTest : public ::testing::Test
{
protected:
    // Setup code (called before each test)
    ShmemPrimitiveTest() : shmHeap("test_shm_heap", 80, 1024), acc(&shmHeap){};

    void SetUp() override
    {
        shmHeap.getLogger()->set_level(spdlog::level::info);
        auto console_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
        shmHeap.getLogger()->sinks().clear();
        shmHeap.getLogger()->sinks().push_back(console_sink);
        // Initialize necessary objects/resources
        shmHeap.create();
        Py_Initialize();
        // Set spdlog sink to a file
    }

    // Teardown code (called after each test)
    void TearDown() override
    {
        // Cleanup objects/resources
        // spdlog::drop_all();
    }

    ShmemHeap shmHeap;
    ShmemAccessor acc;
};

// Core tests
TEST_F(ShmemPrimitiveTest, BasicAssignmentAndMemoryUsage)
{
    size_t expectedMemSize = 0;

    acc = 1;
    expectedMemSize = 24; // min payload is 24
    EXPECT_EQ(shmHeap.briefLayout(), std::vector<size_t>({expectedMemSize, 4096 - expectedMemSize - unitSize * 2}));
    EXPECT_EQ(acc.toString(), "(P:int:1)[1]");

    acc = nullptr;
    EXPECT_EQ(shmHeap.briefLayout(), std::vector<size_t>({4096 - unitSize}));
    EXPECT_EQ(acc.toString(), "nullptr");

    acc = std::vector<float>(10, 1);
    expectedMemSize = 8 + 40; // ShmemPrimitive header 8 bytes + 4*10 + 0 padding
    EXPECT_EQ(shmHeap.briefLayout(), std::vector<size_t>({expectedMemSize, 4096 - expectedMemSize - unitSize * 2}));

    acc = std::vector<int>(10, 1);
    expectedMemSize = 8 + 40; // ShmemPrimitive header 8 bytes + 4*10 + 0 padding
    EXPECT_EQ(shmHeap.briefLayout(), std::vector<size_t>({expectedMemSize, 4096 - expectedMemSize - unitSize * 2}));
    EXPECT_EQ(acc.toString(), "(P:int:10)[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]");

    acc = std::vector<size_t>(10, 1);
    expectedMemSize = 8 + 8 * 10; // ShmemPrimitive header 8 bytes + 8*10 + 0 padding
    EXPECT_EQ(shmHeap.briefLayout(), std::vector<size_t>({expectedMemSize, 4096 - expectedMemSize - unitSize * 2}));
    EXPECT_EQ(acc.toString(), "(P:unsigned long:10)[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]");
    EXPECT_EQ(acc.toString(4), "(P:unsigned long:10)[1, 1, 1, 1, ...]");

    // shmHeap.printShmHeap();
    // std::cout << acc.toString(10) << std::endl;
}

TEST_F(ShmemPrimitiveTest, TypeIdAndLen)
{
    // Single Primitive
    acc = double(7);
    EXPECT_EQ(acc.len(), 1);
    EXPECT_EQ(acc.typeId(), Double);

    // Vector
    acc = std::vector<float>(10, 1);
    EXPECT_EQ(acc.len(), 10);
    EXPECT_EQ(acc.typeId(), Float);

    // String
    acc = "string";
    EXPECT_EQ(acc.len(), 6 + 1);
    EXPECT_EQ(acc.typeId(), Char);
}

TEST_F(ShmemPrimitiveTest, SetAndGetElement)
{
    // Single Primitive
    acc = double(7);
    EXPECT_EQ(acc, 7);
    EXPECT_EQ(acc, std::vector<int>({7}));

    acc = nullptr;
    EXPECT_EQ(acc, nullptr);
    EXPECT_ANY_THROW(acc[0] = 1);

    // Vector
    acc = std::vector<float>(10, 1);

    int a;
    // Assign a multiple element array to a single primitive is not allowed
    EXPECT_ANY_THROW(a = acc);

    acc[0] = 's';
    EXPECT_EQ(acc[0], 's');
    acc[1] = double(11);

    EXPECT_EQ(acc[1], 11);
    acc[2] = 12;
    EXPECT_EQ(acc[2], 12);
    EXPECT_EQ(acc, std::vector<int>({'s', 11, 12, 1, 1, 1, 1, 1, 1, 1}));

    EXPECT_ANY_THROW(acc[10] = "s");

    EXPECT_ANY_THROW(acc[10] = std::vector<int>(10, 1));

    // String
    acc = "string";
    // shmHeap.printShmHeap();
    EXPECT_EQ(acc, "string");
    EXPECT_EQ(acc.len(), 6 + 1);
    EXPECT_EQ(acc.typeId(), Char);
}

TEST_F(ShmemPrimitiveTest, Contains)
{
    acc = std::vector<float>(10, 1);
    acc[0] = 's';
    acc[9] = 'a';

    EXPECT_EQ(acc.contains('s'), true);
    EXPECT_EQ(acc.contains('a'), true);
    EXPECT_EQ(acc.contains('b'), false);
}

TEST_F(ShmemPrimitiveTest, ConvertToPythonObject)
{
    acc = 1;
    EXPECT_EQ(acc.operator py::int_(), 1);
    EXPECT_EQ(acc.operator py::int_(), 1);

    acc = 3.14f;
    EXPECT_FLOAT_EQ(acc.operator py::float_().cast<float>(), 3.14f);

    acc = double(12345678.9123456);
    double val = acc.operator py::float_().cast<double>();
    EXPECT_DOUBLE_EQ(acc.operator py::float_().cast<double>(), 12345678.9123456);
    
    acc = std::vector<float>(10, 1);
    EXPECT_EQ(py::cast<std::string>(py::str(acc.operator py::object())), "[1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]");
    acc = std::vector<int>(10, 1);
    EXPECT_EQ(py::cast<std::string>(py::str(acc.operator py::object())), "[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]");
    acc = std::vector<bool>(10, 1);
    EXPECT_EQ(py::cast<std::string>(py::str(acc.operator py::object())), "[True, True, True, True, True, True, True, True, True, True]");

    EXPECT_ANY_THROW(acc.operator py::int_());
    EXPECT_ANY_THROW(acc.operator py::float_());
    EXPECT_ANY_THROW(acc.operator py::bool_());
    EXPECT_ANY_THROW(acc.operator py::str());
    EXPECT_ANY_THROW(acc.operator py::bytes());

    acc = std::vector<char>(10, 115);
    EXPECT_EQ(py::cast<std::string>(acc.operator py::str()), "sssssssss");
    EXPECT_ANY_THROW(acc.operator py::bytes());

    acc = std::vector<unsigned char>(10, 115);
    EXPECT_EQ(py::cast<std::string>(acc.operator py::bytes()), "ssssssssss");
    EXPECT_ANY_THROW(acc.operator py::str());
}

TEST_F(ShmemPrimitiveTest, AssignWithPythonObject)
{
    py::list l;
    for (int i = 0; i < 10; i++)
    {
        l.append(i);
    }
    acc = l;
    EXPECT_EQ(py::cast<std::string>(py::str(acc.operator py::object())), "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]");
}

File: pytest/ShmemPrimitive_test.py
import pytest
from TypedShmem import ShmemHeap, ShmemAccessor


@pytest.fixture
def shmemPrimitiveTest():
    """Fixture to initialize and cleanup ShmemHeap and ShmemAccessor for each test."""
    shmHeap = ShmemHeap("test_shm_heap", 80, 1024)
    acc = ShmemAccessor(shmHeap)
    shmHeap.setLogLevel(0)  # Set log level to info or appropriate level
    shmHeap.create()  # Create the shared memory heap
    yield shmHeap, acc
    shmHeap.close()  # Clean up resources


def testBasicAssignmentAndMemoryUsage(shmemPrimitiveTest):
    shmHeap, acc = shmemPrimitiveTest
    unitSize = 8  # Assuming unitSize is 8 bytes; adjust based on actual value

    acc.set(1)
    expectedMemSize = 24  # min payload is 24
    assert shmHeap.briefLayout() == [
        expectedMemSize,
        4096 - expectedMemSize - unitSize * 2,
    ]
    assert acc.toString() == "(P:int:1)[1]"

    acc.set(None)
    assert shmHeap.briefLayout() == [4096 - unitSize]
    assert acc.toString() == "nullptr"

    acc.set([1.0] * 10)
    expectedMemSize = 8 + 40  # ShmemPrimitive header 8 bytes + 4*10 + 0 padding
    assert shmHeap.briefLayout() == [
        expectedMemSize,
        4096 - expectedMemSize - unitSize * 2,
    ]

    acc.set([1] * 10)
    expectedMemSize = 8 + 40  # ShmemPrimitive header 8 bytes + 4*10 + 0 padding
    assert shmHeap.briefLayout() == [
        expectedMemSize,
        4096 - expectedMemSize - unitSize * 2,
    ]
    assert acc.toString() == "(P:int:10)[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]"

    # acc.set([1] * 10)
    # expectedMemSize = 8 + 8 * 10  # ShmemPrimitive header 8 bytes + 8*10 + 0 padding
    # assert shmHeap.briefLayout() == [expectedMemSize, 4096 - expectedMemSize - unitSize * 2]
    # assert acc.toString() == "(P:unsigned long:10)[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]"
    # assert acc.toString(4) == "(P:unsigned long:10)[1, 1, 1, 1, ...]"


def testTypeIdAndLen(shmemPrimitiveTest):
    _, acc = shmemPrimitiveTest

    # acc.set(7.0)
    # assert acc.len() == 1
    # assert acc.typeStr() == "double"

    acc.set([1.0] * 10)
    assert acc.len() == 10
    assert acc.typeStr() == "float"

    acc.set("string")
    assert acc.len() == 7
    assert acc.typeStr() == "char"


def testSetAndGetElement(shmemPrimitiveTest):
    _, acc = shmemPrimitiveTest

    acc.set(7.0)
    assert acc.fetch() == 7
    # assert acc.get() == [7]

    acc.set(None)
    assert acc.fetch() is None
    with pytest.raises(Exception):
        acc[0] = 1

    acc.set([1.0] * 10)

    with pytest.raises(Exception):
        acc.get()

    acc[0].set(ord(b"s".decode("ascii")))
    assert acc[0] == ord(b"s".decode("ascii"))
    acc[1].set(11)
    assert acc[1] == 11
    acc[2].set(12)
    assert acc[2] == 12
    assert acc.fetch() == [ord(b"s".decode("ascii")), 11, 12, 1, 1, 1, 1, 1, 1, 1]

    with pytest.raises(Exception):
        acc[10] = ord(b"s".decode("ascii"))

    with pytest.raises(Exception):
        acc[10] = [1] * 10

    acc.set("string")
    assert acc.fetch() == "string"
    assert acc.len() == 7
    assert acc.typeStr() == "char"


# def testContains(shmemPrimitiveTest):
#     _, acc = shmemPrimitiveTest

#     acc.set([1.0] * 10)
#     acc[0] = ord(b"s".decode("ascii"))
#     acc[9] = ord(b"a".decode("ascii"))

#     assert acc.contains(ord(b"s".decode("ascii")))
#     assert acc.contains(ord(b"a".decode("ascii")))
#     assert not acc.contains(ord(b"b".decode("ascii")))


def testConvertToPythonObject(shmemPrimitiveTest):
    _, acc = shmemPrimitiveTest

    acc.set(1)
    assert acc.fetch() == 1

    acc.set(3.14)
    assert acc.fetch() == pytest.approx(3.14)

    acc.set(12345678.9123456)
    assert acc.fetch() == pytest.approx(12345678.9123456)

    acc.set([1.0] * 10)
    assert str(acc.fetch()) == "[1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]"

    acc.set([1] * 10)
    assert str(acc.fetch()) == "[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]"

    acc.set([True] * 10)
    assert (
        str(acc.fetch())
        == "[True, True, True, True, True, True, True, True, True, True]"
    )

    # with pytest.raises(Exception):
    #     acc.fetch()

    acc.set([115] * 10)
    # assert acc.fetch() == "ssssssssss"
    # assert str(acc.fetch()) == "[115, 115, 115, 115, 115, 115, 115, 115, 115, 115]"


def testAssignWithPythonObject(shmemPrimitiveTest):
    _, acc = shmemPrimitiveTest

    acc.set(list(range(10)))
    assert str(acc.fetch()) == "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]"

    acc.set([0, 1, 2, 3, 4, 100, 6, 7, 8, 9])
    assert str(acc.fetch()) == "[0, 1, 2, 3, 4, 100, 6, 7, 8, 9]"

    acc.set(True)
    assert acc.fetch() is True

    acc.set(3)
    assert acc.fetch() == 3

    acc.set(3.14)
    assert acc.fetch() == pytest.approx(3.14)

    acc.set("sample")
    assert acc.fetch() == "sample"

    # with pytest.raises(Exception):
    #     acc.fetch()

    acc.set(b"sample")
    # with pytest.raises(Exception):
    #     acc.fetch()

    assert acc.fetch() == [
        ord(b"s".decode("ascii")),
        ord(b"a".decode("ascii")),
        ord(b"m".decode("ascii")),
        ord(b"p".decode("ascii")),
        ord(b"l".decode("ascii")),
        ord(b"e".decode("ascii")),
    ]
    print(acc.fetch())


if __name__ == "__main__":
    pytest.main(["-v", "pytest/ShmemPrimitive_test.py"])
